tithemi/assignment4/islands.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <array>
+#include <vector>
+#include <set>
+#include <iostream>
+
+template<size_t Rows, size_t Columns>
+size_t count_islands(const std::array<std::array<bool, Columns>, Rows>& map)
+{
+    size_t counter = -1;
+
+    std::vector<size_t*> islands;
+    std::array<std::array<size_t*, Columns>, Rows> island_ptr{};
+
+    for (size_t i = 0; i < Rows; ++i)
+    {
+        for (size_t j = 0; j < Columns; ++j)
+        {
+            if (!map[i][j])
+                continue;
+
+            size_t* top = i ? island_ptr[i - 1][j] : nullptr;
+            size_t* left = j ? island_ptr[i][j - 1] : nullptr;
+
+            if (top)
+            {
+                if (left)
+                {
+                    *top = *left;
+                    island_ptr[i][j] = left;
+                }
+                else
+                {
+                    island_ptr[i][j] = top;
+                }
+            }
+            else
+            {
+                if (left)
+                {
+                    island_ptr[i][j] = left;
+                }
+                else
+                {
+                    islands.push_back(new size_t(++counter));

```

Comment:
Maybe I'm misunderstanding your algorithm, but why do you need to store a counter vector? You only ever use the last element, isn't it the same as just keeping the single counter value?